crates/ty_python_semantic/src/types.rs

```diff
@@ -5015,8 +5016,9 @@ impl<'db> Type<'db> {
                 SpecialFormType::Callable => Ok(CallableType::unknown(db)),
 
                 SpecialFormType::TypingSelf => {
+                    let module = parsed_module(db.upcast(), scope_id.file(db)).load(db.upcast());
                     let index = semantic_index(db, scope_id.file(db));
```

Comment:
@carljm it seems problematic to me that this `Type` method access the AST. Are there some guardrails in place that I can't see from the method's signature that prevent this from creating cross-module dependencies? Maybe there is because we pass in scope? Should we make this method `pub(crate)`

Reply:
I think this case is currently fine in practice, because we pass in `scope_id`, and this is always a `scope_id` from the file we are currently inferring, never cross-file. You can see this from the fact that every call site of `Type::in_type_expression` in `TypeInferenceBuilder` passes in `self.scope()` for the scope argument.

It would definitely be fine to make this method `pub(crate)`, though I don't know how much that would really help prevent a possible future mis-use.

Not sure what else we could do here...

Reply:
https://github.com/astral-sh/ruff/pull/18488 -- open to additional suggestions